.github/workflows/deploy_mcp.yml

```diff
@@ -0,0 +1,151 @@
+          gcloud auth activate-service-account --key-file=sa-key.json
+
+      # Configure yaml files to contain relevant project info
+      - name: Configure Files
+        run: |
+          cd GCP
+          echo "name: ${{env.GCP_PROJECT_ID}}_app_bucket" > project.yml
+          cat project.yml
+          echo "project_id: &project_id ${{env.GCP_PROJECT_ID}}" | cat - gcp_ae.yml > temp && mv temp gcp_ae.yml
+          echo "app_name: &app_name ${{env.GCP_PROJECT_ID}}" | cat - gcp_ae.yml > temp && mv temp gcp_ae.yml
+          cat gcp_ae.yml
+          echo "image_uri: &image_uri europe-west2-docker.pkg.dev/${{env.GCP_PROJECT_ID}}/app-repo/helloworld:latest" | cat - gcp_cloudrun.yml > temp && mv temp gcp_cloudrun.yml
+          echo "project_id: ${{env.GCP_PROJECT_ID}}" | cat - gcp_cloudrun.yml > temp && mv temp gcp_cloudrun.yml
+          cat gcp_cloudrun.yml
+
+      # Install kubectl and create gke cluster
+      - name: setup kubectl
+        run: |
+          gcloud components install kubectl
+          gcloud container clusters create mcpdeploytest-cluster --num-nodes=1
+          gcloud container clusters get-credentials mcpdeploytest-cluster
+
+      # Setup main.tf, deploy and destroy k8s
+      - name: deploy k8s
+        run: |
+          cd k8s
+          mkdir terraform
+          cd terraform
+          touch main.tf
+          {
+            echo "terraform {"
+            echo "  backend \"gcs\"{"
+            echo "    bucket = \"${{env.GCP_PROJECT_ID}}_bucket\""
+            echo "    prefix = \"tf-state-files\""
+            echo "  }"
+            echo "}"
+            echo "module k8s {"
+            echo "  source = \"../../../../../mcp\""
+            echo "}"
+          } > main.tf
+          cat main.tf
+          terraform init
+          terraform refresh
+          terraform plan -out="./plan.tfplan"
+          terraform apply plan.tfplan
+          terraform plan -destroy -out="./destroy.tfplan"
+          terraform apply destroy.tfplan
+
+      # Setup main.tf and deploy gcp modules. Also plan destroy but don't apply due to app engine issues.
+      - name: Deploy GCP modules
+        run: |
+          cd GCP
+          mkdir terraform
+          cd terraform
+          touch main.tf
+          {
+            echo "terraform {"
+            echo "  backend \"gcs\"{"
+            echo "    bucket = \"${{env.GCP_PROJECT_ID}}_bucket\""
+            echo "    prefix = \"tf-state-files\""
+            echo "  }"
+            echo "}"
+            echo "module gcp {"
+            echo "  source = \"../../../../../mcp\""
+            echo "}"
+          } > main.tf
+          cat main.tf
+          terraform init
+          terraform refresh
+          terraform plan -out="./plan.tfplan"
+          terraform apply plan.tfplan
+          terraform plan -destroy -out="./destroy.tfplan"
```

Comment:
If you had a repository which container a Terraform deployment module acting as a template for how to perform a deployment, would that make all of these scripted tasks much simpler?

For example:
1. task to git pull that repo
2. task to make any use-case modifications to that code based on Action variables
3. apply terraform
4. destroy terraform

Reply:
Most of the scripted tasks are gcloud setup and configuring the files to include the variables for the project. The setup for main.tf is long because I thought it was clearer to write the whole file rather than try to add the bucket name to a line in the middle of a file. 
I will look into whether it could be simpler though

Reply:
I feel that using a seperate repository will lead to two different areas needing to be maintained. I may be misunderstanding but would this be a repository containing `main.tf` templates? 
I had the script create `main.tf` because the line that would need modifying is `bucket = "${{env.GCP_PROJECT_ID}}_bucket"` but that would require inserting a line in the middle of a files, or appending to the end of a found string. I felt creating the whole file was clearer.
I can remove `terraform refresh`, but I did `terraform plan` then `terraform apply plan` so that there would be an output of the plan used for debugging purposes. `terraform apply --auto-approve` doesn't display the plan. The same applies with the destroy plan, but the gcp module does not apply the destoy plan to avoid the error destroying app engine.

Reply:
> I feel that using a seperate repository will lead to two different areas needing to be maintained. 
Doesn't need to be a separate repository. Also, consider what the next task after Github Actions (TeamCity) is and if we add a new adapter that needs testing.

Potentially we could also use the same test data (main.tf) to test [other applications](https://github.com/mesoform/example-ecat-deployment-code/blob/master/gcp/project.tf)

> I had the script create main.tf because the line that would need modifying is bucket = "${{env.GCP_PROJECT_ID}}_bucket"
A single `sed` line will [change out any known placeholders](https://github.com/mesoform/Multi-Cloud-Platform/blob/d8d4f61435ae29ff4d8f67fb15478cc730b6eb67/monitoring/moniadm.sh#L216).

> I can remove terraform refresh
What's the default refresh action of `terraform plan`?

> I did `terraform plan` then `terraform apply`
That's a good call

Reply:
I have updated the file setup to use sed to replace placeholder. The `main.tf` files are included in their relevant directories. 
>What's the default refresh action of `terraform plan`  

`terraform plan` refreshes unless explicitly told not to. So have removed this line.